### DISCUSSION TITLE
Exposing user impersonation as auth function

### DISCUSSION THREAD

**Discussion by devidevio** [supabase/supabase#19668] [Feature Requests]
It would be great to see the feature of [user impersonation](https://supabase.com/blog/studio-introducing-assistant#user-impersonation) as a auth helper function within supabase, f. e. `auth.impersonate(userId);` for the lifetime of a transaction.

The code does not seem so complex, just the logic to obtain a `session_id` and `sub` could be integrated in this flow, example code from the Supabase Studio:
```
    select set_config('role', 'authenticated', true),
           set_config('request.jwt.claims', '{"aal":"aal1","amr":[{"method":"password","timestamp":1702465165}],"app_metadata":{"provider":"email","providers":["email"]},"aud":"authenticated","email":"my@email.com","exp":1702468765,"iat":1702465165,"iss":"https://projectref.supabase.co/auth/v1","phone":null,"role":"authenticated","session_id":"UUID_PLACEHOLDER","sub":"UUID_PLACEHOLDER","user_metadata":{}}', true),
           set_config('request.method', 'POST', true),
           set_config('request.path', '/impersonation-example-request-path', true),
           set_config('request.headers', '{"accept": "*/*"}', true);

    -- If the users sql returns no rows, pg-meta will
    -- fallback to returning the result of the impersonation sql.
    select 1 as "ROLE_IMPERSONATION_NO_RESULTS";

    select * from public.invoices order by id desc nulls last limit 100 offset 0;
```

Thanks in advance.

**Comment by richardfwashington** [comment]
I'd love to see this too.  It feels like an important, but totally missing, feature. We are currently having to hack a solution together, which doesn't feel good.  Other Auth providers (such as Clerk) do deliver this feature. https://clerk.com/docs/users/user-impersonation

It is important that some of our users are able to log in as and act as others in some cases.
